File: 026_Is Subsequence.cpp
/*
Is Subsequence

Check if s is a subsequence of t.

Strategy: Two-pointer scan over t to match s.
*/

#include <string>
#include <iostream>
using namespace std;

bool isSubsequence(const string& s, const string& t) {
    int i = 0, j = 0;
    while (i < (int)s.size() && j < (int)t.size()) {
        if (s[i] == t[j]) ++i;
        ++j;
    }
    return i == (int)s.size();
}

int main() {
    cout << boolalpha;
    cout << "IsSub 1: " << isSubsequence("abc", "ahbgdc") << '\n';
    cout << "IsSub 2: " << isSubsequence("axc", "ahbgdc") << '\n';
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(1)
*/

File: 026_Is Subsequence.py
"""
Is Subsequence

Check if s is a subsequence of t.
"""

def is_subsequence(s: str, t: str) -> bool:
    it = iter(t)
    for ch in s:
        found = False
        for x in it:
            if x == ch:
                found = True
                break
        if not found:
            return False
    return True


if __name__ == '__main__':
    print('IsSub 1:', is_subsequence('abc', 'ahbgdc'))
    print('IsSub 2:', is_subsequence('axc', 'ahbgdc'))

# Time Complexity: O(n)
# Space Complexity: O(1)
